common/src/reserved.rs

```diff
@@ -338,4 +409,191 @@ mod tests {
+        let new_state_validator_set = new_state.get_validator_set();
+        let new_state_governance_set = new_state.get_governance_set();
+
+        assert_eq!(
+            new_state.members[1].governance_delegatee.as_ref().unwrap(),
+            &delegatee.name
+        );
+
+        assert_eq!(
+            new_state.members[1].consensus_delegatee.as_ref().unwrap(),
+            &delegatee.name
+        );
+
+        assert_eq!(
+            new_state_validator_set
+                .unwrap()
+                .into_iter()
+                .find(|v| v.0 == delegatee.public_key)
+                .unwrap()
+                .1,
+            3
+        );
+
+        assert_eq!(
+            new_state_governance_set
+                .unwrap()
+                .into_iter()
+                .find(|v| v.0 == delegatee.public_key)
+                .unwrap()
+                .1,
+            3
+        );
+    }
+
+    #[test]
+    fn test_apply_delegate_on_consensus_success() {
+        // given
+        setup_test();
+        let (mut state, keys) = generate_delegated_genesis(3);
+        let delegator_public_key = keys[0].0.clone();
+        let delegator_private_key = keys[0].1.clone();
+        let delegatee_public_key = keys[1].0.clone();
+
+        let delegatee = state.members[2].clone();
+
+        // when
+        let data: DelegationTransactionData = DelegationTransactionData {
+            delegator: delegator_public_key.clone(),
+            delegatee: delegatee_public_key.clone(),
+            governance: false,
+            block_height: 0u64,
+        };
+        let proof = TypedSignature::sign(&data, &delegator_private_key).unwrap();
+
+        let tx = TxDelegate {
+            delegator: delegator_public_key,
+            delegatee: delegatee_public_key,
+            governance: false,
+            proof,
+            timestamp: 0,
+        };
+
+        let new_state = state.apply_delegate(&tx);
+
+        // then
+        assert!(new_state.is_ok());
+        let new_state = new_state.unwrap();
+        let new_state_validator_set = new_state.get_validator_set();
+        let new_state_governance_set = new_state.get_governance_set();
+
+        assert_eq!(
+            new_state.members[1].consensus_delegatee.as_ref().unwrap(),
+            &delegatee.name
+        );
+
+        assert_eq!(new_state.members[1].governance_delegatee, None);
+
+        assert_eq!(
+            new_state_validator_set
+                .unwrap()
+                .into_iter()
+                .find(|v| v.0 == delegatee.public_key)
+                .unwrap()
+                .1,
+            3
+        );
+
+        assert_eq!(
+            new_state_governance_set
+                .unwrap()
+                .into_iter()
+                .find(|v| v.0 == delegatee.public_key)
+                .unwrap()
+                .1,
+            2
+        );
+    }
+
+    #[test]
+    fn test_apply_undelegate_success() {
```

Comment:
1. `generated_delegated_genesis` to init delegation
2. `undelegation` logic
3. assert `get_validator_set` & `get_governance_set`